diffrax/_solver/srk.py

```diff
@@ -0,0 +1,649 @@
+                    w_kg_sum,
+                    *levy_sum_list,
+                )
+
+            # compute Σ_{i=1}^{j-1} a_j_i h_kf_i
+            _drift_result = sum_prev_stages(_h_kfs, a_j)
+
+            # z_j = y_0 + h (Σ_{i=1}^{j-1} a_j_i k_i) + g * (a_w_j * ΔW + cH_j * ΔH)
+            z_j = (y0**ω + _drift_result**ω + _diffusion_result**ω).ω
+
+            def compute_and_insert_kf_j(_h_kfs_in):
+                h_kf_j = drift.vf_prod(t0 + c_j * h, z_j, args, h)
+                return insert_jth_stage(_h_kfs_in, h_kf_j, j)
+
+            if self.tableau.ignore_stage_f is None:
+                _h_kfs = compute_and_insert_kf_j(_h_kfs)
+            else:
+                drift_pred = jnp.logical_not(ignore_stage_f[j])
+                _h_kfs = lax.cond(
+                    eqxi.nonbatchable(drift_pred),
+                    compute_and_insert_kf_j,
+                    lambda what: what,
+                    _h_kfs,
+                )
+
+            if additive_noise:
+                return (_h_kfs, None, None), None
+
+            def compute_and_insert_kg_j(_w_kgs_in, _levylist_kgs_in):
+                _w_kg_j = diffusion.vf_prod(t0 + c_j * h, z_j, args, w)
+                new_w_kgs = insert_jth_stage(_w_kgs_in, _w_kg_j, j)
+
+                _levylist_kg_j = [
+                    diffusion.vf_prod(t0 + c_j * h, z_j, args, levy)
+                    for levy in levy_areas
+                ]
+                new_levylist_kgs = insert_jth_stage(_levylist_kgs_in, _levylist_kg_j, j)
+                return new_w_kgs, new_levylist_kgs
+
+            if self.tableau.ignore_stage_g is None:
+                _w_kgs, _levylist_kgs = compute_and_insert_kg_j(_w_kgs, _levylist_kgs)
+            else:
+                diffusion_pred = jnp.logical_not(ignore_stage_g[j])
+                _w_kgs, _levylist_kgs = lax.cond(
+                    eqxi.nonbatchable(diffusion_pred),
+                    compute_and_insert_kg_j,
+                    lambda x, y: (x, y),
+                    _w_kgs,
+                    _levylist_kgs,
+                )
+
+            return (_h_kfs, _w_kgs, _levylist_kgs), None
+
+        scan_out = eqxi.scan(
+            stage,
+            carry,
+            scan_inputs,
+            len(b_sol),
+            buffers=lambda x: x,
+            kind="checkpointed",
+            checkpoints="all",
+        )
+
+        if additive_noise:
+            # output of lax.scan is ((num_stages, _h_kfs), None)
+            (h_kfs, _, _), _ = scan_out
+            diffusion_result = jtu.tree_map(
+                add_levy_to_w(b_w, *b_levy_list), w_kg, *levylist_kgs
+            )
+
+            # In the additive noise case (i.e. when g is independent of y),
+            # we still need a correction term in case the diffusion vector field
+            # g depends on t. This term is of the form $(g1 - g0) * (0.5*W_n - H_n)$.
+            g1 = diffusion.vf(t1, y0, args)
```

Comment:
I still need to examine the computation graph here a little more closely, but something to be aware of: each *call site* (i.e. the number of times it is literally written out in the code) is compiled independently. This means that if you make lots of evaluations of e.g. `diffusion.vf`, then compilation times can increase quite dramatically.

This is actually the reason we use an `eqxi.scan` here, and not a normal Python `for` loop!

Sometimes we really do need additional call sites, of course, just a heads-up to try and minimise the number.